[Bug fix]: Unscheduable host-model VMI alert is not triggered

iholder-redhat opened the pull request:
**What this PR does / why we need it**:
This is PR to fix a bug this PR: https://github.com/kubevirt/kubevirt/pull/6773. 

This PR fixes a bug that caused alert for unschedulable host-model VMIs to trigger.
Two bugs are now fixed:
- In `alertIfHostModelIsUnschedulable()` function, which scans all nodes to check whether one or more nodes support the VMI host model and cpu required features, the source node is now skipped. If the source node is not skipped, it is always found to be suitable, obviously.
- In `isNodeSuitableForHostModelMigration()` function, iterate on Pod's `NodeSelector`, not `Labels`.

In addition, the PR includes several improvements:
- Introduces unit and functional test
- Moves logic to `handlePendingPodTimeout()` which already deals with unschedulable migration target pods
- Refactoring and other small improvements

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Bug fix: Unscheduable host-model VMI alert is now properly triggered
```

iholder-redhat commented:
/cc @vladikr

xpivarc commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -1397,6 +1391,18 @@ func (c *MigrationController) getNodeForVMI(vmi *virtv1.VirtualMachineInstance)
 	return node, nil
 }
 
+func (c *MigrationController) triggerAlertForHostModelUnschedulable(vmi *virtv1.VirtualMachineInstance, targetPod *k8sv1.Pod) {
```
I would suggest naming this `alertIfHostModelIsUnschedulable` or similarly. As of now, I think this will alert always unless I look to the code.

iholder-redhat replied:
Nice! done

xpivarc commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -1391,12 +1391,14 @@ func (c *MigrationController) getNodeForVMI(vmi *virtv1.VirtualMachineInstance)
 }
 
 func (c *MigrationController) triggerAlertForHostModelUnschedulable(vmi *virtv1.VirtualMachineInstance, targetPod *k8sv1.Pod) {
+	const warningMessage = "Migration cannot proceed since no node is suitable to run the required CPU model / required features"
```
Is this even correct? 
We iterate through all nodes but we fire the alert for each node that is not suitable.

iholder-redhat replied:
Thanks for catching that! You're absolutely right!

xpivarc commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -922,6 +920,7 @@ func (c *MigrationController) handlePendingPodTimeout(migration *virtv1.VirtualM
 	secondsSpentPending := timeSinceCreationSeconds(&pod.ObjectMeta)
 
 	if isPodPendingUnschedulable(pod) {
+		c.triggerAlertForHostModelUnschedulable(vmi, pod)
```
I am not sure I understand the reasoning of https://github.com/kubevirt/kubevirt/pull/6918/commits/f5bb8e9a1b5182a03b57946e7c0bc9df921db945 . Why doesn't the execution continue to `updateStatus`?

iholder-redhat replied:
For example, if the timeout exceeded we would delete target pod and have an alert saying VMI cannot be scheduled, then the logic on `updateStatus()` becomes irrelevant. The right behavior here is that even if timeout exceeded and target pod is deleted we want the more precise alert (about host-model) and not the more general alert (about not being able to schedule)

xpivarc reviewed (changes requested):
Could we add a test that would verify this?

iholder-redhat commented:
@xpivarc 
>Could we add a test that would verify this?

Done.

iholder-redhat commented:
@xpivarc raised some valid points. Thank you for that. I'm converting this PR to draft until I figure out what's going on.

dshchedr commented on pkg/virt-controller/watch/migration_test.go:
```diff
@@ -1180,6 +1180,44 @@ var _ = Describe("Migration watcher", func() {
+
+	Context("Migration of host-model VMI", func() {
+
+		It("should trigget alert when no node supports host-model", func() {
```
just typo: trigge**r**

iholder-redhat commented:
This PR is ready to review again. I've added a functional test which verifies that.
Please see updated PR info.

@xpivarc Thanks very much for your awesome review!
@dshchedr Typo's fixed, thanks!
@vladikr FYI

vladikr commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -1426,15 +1446,11 @@ func prepareNodeSelectorForHostCpuModel(node *k8sv1.Node, pod *k8sv1.Pod) error
 	}
 
-	for key, value := range pod.Labels {
+	for key, value := range pod.Spec.NodeSelector {
```
uhh, how did this work before?

iholder-redhat replied:
It didn't :smiling_face_with_tear:

vladikr commented on tests/migration_test.go:
```diff
@@ -2300,6 +2302,101 @@ var _ = Describe("[Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][leve
+
+					By("Expecting for an alert to be triggered")
+					Eventually(func() []k8sv1.Event {
+						//var hostModelEvents []k8sv1.Event
```
left over

iholder-redhat replied:
Thanks!

vladikr commented on tests/migration_test.go:
```diff
@@ -2300,6 +2302,101 @@ var _ = Describe("[Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][leve
 				expectFeatureToBeSupportedOnNode(newNode, requiredFeatures)
 			})
+
+			Context("[Serial]Should trigger event", func() {
```
:+1: great!

vladikr reviewed (commented):
Looks great! Thanks.
Perhaps we should log and add to the event the translation of the host-model?

iholder-redhat commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -1397,6 +1390,33 @@ func (c *MigrationController) getNodeForVMI(vmi *virtv1.VirtualMachineInstance)
+			}
+
+			if isNodeSuitableForHostModelMigration(node, targetPod) {
+				log.Log.Object(vmi).Infof("Node %s is suitable to run vmi %s host model cpu mode (more nodes may fit as well)", node.Name, vmi.Name)
```
By "translation" you mean the host CPU model which we're trying to match to?
I thought of that, the thing is that it can be very verbose, as CPU required features are also crucial here and there could be many of them. On the other hand this surely can be helpful for debugging.

No problem, I'll add that

iholder-redhat commented:
I fixed a small bug I had (uninitialized map... :facepalm:), looking good now :+1: 
/retest

vladikr commented:
Thanks, look good to me
/approve

@xpivarc could you please take another look?

xpivarc reviewed (changes requested):
Please squash the commits so future us have an easier time with backporting and finding the origin of changes :)

iholder-redhat commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -1407,6 +1400,43 @@ func (c *MigrationController) getNodeForVMI(vmi *virtv1.VirtualMachineInstance)
 }
 
+func (c *MigrationController) alertIfHostModelIsUnschedulable(vmi *virtv1.VirtualMachineInstance, targetPod *k8sv1.Pod) {
+	const warningMsgPattern = "Migration cannot proceed since no node is suitable to run the required CPU model / required features: %v"
```
I tend to disagree :)

This is relevant to a very specific scope, so I don't see a reason to get it out to a broader scope. In general I think this should be the approach - variables need to be declared at the most inner scope they're relevant to. This way anyone who reads the code has a clear idea of what this variable affects (and what it can't affect).

iholder-redhat commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -1407,6 +1400,43 @@ func (c *MigrationController) getNodeForVMI(vmi *virtv1.VirtualMachineInstance)
+	const warningMsgPattern = "Migration cannot proceed since no node is suitable to run the required CPU model / required features: %v"
+	fittingNodeFound := false
+
+	if cpu := vmi.Spec.Domain.CPU; cpu == nil || cpu.Model != virtv1.CPUModeHostModel || !isPodPending(targetPod) {
```
Because if the scenario is that the Pod cannot be scheduled due to no node supporting its model (or in other words - no node has labels that fit to its nodeSelector) it will be in pending phase. If, for example, the pod is already scheduled or scheduling - there is no point of making these checks.

iholder-redhat commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -1434,21 +1464,15 @@ func prepareNodeSelectorForHostCpuModel(node *k8sv1.Node, pod *k8sv1.Pod) error
-				return false
-			}
+	for key, value := range requiredNodeLabels {
+		if !nodeHasLabel(key, value) {
```
Yup, I agree. Changed it. thanks!

iholder-redhat commented:
@xpivarc 
> Please squash the commits so future us have an easier time with backporting and finding the origin of changes :)
Done :) thanks for your review!

iholder-redhat commented:
/test pull-kubevirt-unit-test

xpivarc reviewed (commented):
Thanks for squashing the commits but I believe we still can improve here. Few examples:

https://github.com/kubevirt/kubevirt/pull/6918/commits/c49e16b76a85eef4104519f68b5027999b7a04cb is highlighting **Add functional test to ensure one alert is triggered** but I personally feel that fixing 2 bugs mentioned below is THE thing to highlight. Summary as **Fix NoSuitableNodesForHostModelMigration triggering logic ** would be more informative for future us. Of course, I am ok with including the test within the change and I think we can also agree it's definitely important.

https://github.com/kubevirt/kubevirt/pull/6918/commits/323710cbb7bd80153fd96db6d36035876137c708 includes **Change triggerAlertForHostModelUnschedulable() to alertIfHostModelIsUnschedulable()**. I personally think we should include this in the commit that actually refactors the functions and is defining the name for the function. Otherwise, it's confusing to me why it is included in one changeset(PR). I would even add the change which is moving the defined function as it is still only refactoring(no functional change - only ordering is changing afaik) and therefore doesn't need an explanation on its own.

WDYT?

iholder-redhat commented:
@xpivarc 
I've changed the commits in the way you suggested.

> We now check this in isPodPendingUnschedulable. We had to do it before.

Yup, you're right, changed that, thanks!

> I personally don't think it's needed. In this case, it's only a pattern that is later used for event message. It's constant and as written here it will be created at every call.
>
> I definitely don't block you on this.

I made a literal string in `Sprintf` instead of the const. Performance-wise this is exactly the same (constants are created at compilation time, same as string literals). But since the const variable was only used once I guess it doesn't really matter if it's there or not. Tell me what you think.

iholder-redhat commented:
/retest

iholder-redhat commented:
/retest

iholder-redhat commented:
/retest

iholder-redhat commented:
/cherrypick release-0.48
